File: backend/app/models/release_validation.py
from datetime import datetime, timezone
from enum import StrEnum

from bson import ObjectId
from pydantic import BaseModel, Field


class ReleaseValidationStatus(StrEnum):
    PENDING = "pending"
    RUNNING = "running"
    COMPUTING = "computing"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"
    # 3.5 Approval workflow
    AWAITING_APPROVAL = "awaiting_approval"
    # 9B.4 Explicit approved/rejected terminal states
    APPROVED = "approved"
    REJECTED = "rejected"


class ReleaseValidationContext(BaseModel):
    prd_text: str = ""
    notes: str = ""


class ReleaseValidation(BaseModel):
    id: ObjectId | None = Field(default=None, alias="_id")
    org_id: ObjectId
    project_id: ObjectId
    project_name: str = ""
    title: str = ""
    triggered_by: ObjectId
    status: ReleaseValidationStatus = ReleaseValidationStatus.PENDING
    batch_id: str = ""
    total_runs: int = 0
    completed_runs: int = 0
    passed_runs: int = 0
    failed_runs: int = 0
    error_runs: int = 0
    context: ReleaseValidationContext = Field(default_factory=ReleaseValidationContext)
    confidence_score: int | None = None
    confidence_grade: str | None = None
    recommendation: str | None = None
    report: dict = Field(default_factory=dict)
    score_version: str | None = None
    trend: str | None = None
    risk_delta: int | None = None
    historical_confidence: int | None = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    completed_at: datetime | None = None

    # 1.4 Score freeze — snapshotted at decision time, never recomputed
    final_score_at_decision: int | None = None
    decision_at: datetime | None = None
    score_frozen: bool = False

    # 1.2 Override audit trail
    override_by: ObjectId | None = None
    override_reason: str | None = None
    override_at: datetime | None = None
    override_type: str | None = None  # "ship_anyway" | "acknowledge_risk"
    original_decision: str | None = None

    # 2.4 Production outcome integration
    outcome: str | None = None  # "production_passed" | "production_failed" | "rolled_back"
    outcome_recorded_at: datetime | None = None
    outcome_notes: str | None = None

    # 2.5 Validation SLA — set at creation; 0 means no limit
    sla_minutes: int = 0
    sla_breached: bool = False
    sla_breached_at: datetime | None = None

    # 3.1 Change-aware test selection metadata
    changed_areas: list[str] = Field(default_factory=list)
    targeted_selection: bool = False  # True when test cases were filtered by changed_areas

    # Task-level error (set when status = "failed")
    error_message: str | None = None

    # 2.3 Prediction engine validation
    predicted_score_pre_run: int | None = None   # predicted score captured at validation creation
    prediction_accuracy: int | None = None        # actual - predicted (positive = better than expected)

    # 3.3 Webhook delivery tracking
    webhook_delivery_status: str | None = None   # "pending" | "delivered" | "failed"
    webhook_last_attempt_at: datetime | None = None

    # 4.3 Outcome recording nudge
    outcome_reminder_sent_at: datetime | None = None

    # 5.1 AI adjustment audit log
    ai_adjustment_detail: dict | None = None

    # 7C.2 Incident signal attribution (only set when outcome=production_failed)
    incident_signal_attribution: list[dict] | None = None

    # 8.1 Score degradation — set when one or more scoring sub-engines failed
    score_degraded: bool = False
    degraded_engines: list[str] = Field(default_factory=list)

    # 3.5 Approval workflow
    approval_required: bool = False
    approved_by: ObjectId | None = None
    approved_at: datetime | None = None
    rejected_by: ObjectId | None = None
    rejected_at: datetime | None = None
    rejection_reason: str | None = None

    model_config = {
        "populate_by_name": True,
        "arbitrary_types_allowed": True,
        "json_encoders": {ObjectId: str},
    }

File: frontend/src/types/release-validation.ts
export type ReleaseValidationStatus =
  | "pending"
  | "running"
  | "computing"
  | "completed"
  | "failed"
  | "cancelled"
  | "awaiting_approval"
  | "approved"
  | "rejected";

export interface ReleaseValidation {
  id: string;
  org_id: string;
  project_id: string;
  project_name: string;
  title: string;
  status: ReleaseValidationStatus;
  batch_id: string;
  total_runs: number;
  completed_runs: number;
  passed_runs: number;
  failed_runs: number;
  error_runs: number;
  context: { prd_text: string; notes: string };
  confidence_score: number | null;
  confidence_grade: string | null;
  recommendation: string | null;
  report: ReleaseReport;
  // V1/V2/V3 scoring metadata
  score_version: string | null;
  trend: "up" | "down" | "stable" | null;
  risk_delta: number | null;
  historical_confidence: number | null;
  created_at: string;
  updated_at: string;
  completed_at: string | null;
  // 1.4 Score freeze
  final_score_at_decision: number | null;
  decision_at: string | null;
  score_frozen: boolean;
  // 1.2 Override audit trail
  override_by: string | null;
  override_reason: string | null;
  override_at: string | null;
  override_type: "ship_anyway" | "acknowledge_risk" | null;
  original_decision: string | null;
  // 2.4 Production outcome
  outcome: "production_passed" | "production_failed" | "rolled_back" | null;
  outcome_recorded_at: string | null;
  outcome_notes: string | null;
  // 2.5 Validation SLA
  sla_minutes: number;
  sla_breached: boolean;
  sla_breached_at: string | null;
  // 3.1 Change-aware selection
  changed_areas: string[];
  targeted_selection: boolean;
  // 3.5 Approval workflow
  approval_required: boolean;
  approved_by: string | null;
  approved_at: string | null;
  rejected_by: string | null;
  rejected_at: string | null;
  rejection_reason: string | null;
  // Task error
  error_message: string | null;
  // 5.1 AI adjustment audit
  ai_adjustment_detail: {
    model: string;
    has_prd: boolean;
    raw_adjustment: number;
    clamped_adjustment: number;
    ai_confidence: number;
    input_signals: { pre_ai_score: number; pass_rate: number; instability_index: number; coverage_score: number; trend: string };
    insights: string[];
    risk_explanations: string[];
  } | null;
  // 2.3 Prediction accuracy
  predicted_score_pre_run: number | null;
  prediction_accuracy: number | null;
  // 3.3 Webhook delivery
  webhook_delivery_status: "pending" | "delivered" | "failed" | null;
  webhook_last_attempt_at: string | null;
  // 8.1 Score degradation
  score_degraded: boolean;
  degraded_engines: string[];
  // 7C.2 Incident signal attribution
  incident_signal_attribution: Array<{
    signal: string;
    value: number;
    org_mean: number;
    z_score: number;
    direction: "high" | "low";
  }> | null;
}

export interface ReleaseReport {
  confidence_score?: number;
  confidence_grade?: string;
  recommendation?: string;
  recommendation_reasons?: string[];
  signals?: ReleaseSignal[];
  root_causes?: RootCause[];
  blockers?: string[];
  batch_summary?: {
    total: number;
    passed: number;
    failed: number;
    error: number;
    pass_rate: number;
  };
  per_run_results?: PerRunResult[];
  feature_risks?: FeatureRisk[];
  flaky_tests?: FlakyTest[];
  timing_anomalies?: TimingAnomaly[];
  // Scoring metadata
  score_version?: string;
  base_score?: number;
  adjusted_score?: number;
  healing_penalty?: number;
  instability_index?: number;
  instability_penalty?: number;
  instability_components?: { retry_rate: number; healing_rate: number; behavior_failure_rate: number; resolution_retry_rate: number };
  coverage_score?: number;
  coverage_penalty?: number;
  risk_adjustment?: number;
  high_risk_steps?: { step_number: number; risk_weight: number; detail: string }[];
  ai_adjustment?: number;
  ai_confidence?: number;
  ai_insights?: string[];
  risk_explanations?: string[];
  predicted_failure_probability?: number;
  calibration_confidence?: number;
  score_confidence?: number;
  data_quality?: "LOW" | "MEDIUM" | "HIGH";
  freshness?: "HIGH" | "MEDIUM" | "LOW";
  decayed_score?: number;
  hours_since_run?: number;
  trend?: string;
  historical_confidence?: number;
  risk_delta?: number;
  trajectory?: { scores: number[]; labels: string[] };
  anomalies?: { type: string; metric: string; severity: string; deviation: number; detail: string }[];
  delta?: { score_delta: number | null; pass_rate_delta: number | null; reasons: string[]; has_previous: boolean };
  ai_summary?: {
    executive_summary: string;
    key_findings: string[];
    risk_areas: string[];
    next_steps: string[];
    ai_generated: boolean;
  };
  // 2.1 Actionable failure surface
  actionable_steps?: ActionableStep[];
  // Hard gate results
  gate_blocks?: string[];
  gate_warnings?: string[];
  insufficient_data?: boolean;
  // 2.1 Assertion coverage
  no_assertion_tests?: string[];
  // 6.1 Route coverage
  route_coverage?: {
    batch_routes: string[];
    known_routes: string[];
    coverage_pct: number | null;
  };
  // 7B.4 PRD requirement traceability
  requirement_coverage?: Array<{
    requirement: string;
    covered: boolean;
    evidence: string | null;
  }>;
}

export interface ActionableStep {
  priority: "critical" | "high" | "medium";
  title: string;
  detail: string;
  action: string;
}

export interface ReleaseSignal {
  name: string;
  category: string;
  severity: "good" | "warning" | "critical";
  score_contribution: number;
  detail: string;
  affected_tests: string[];
}

export interface RootCause {
  rank: number;
  description: string;
  impact_score: number;
  affected_tests: string[];
  affected_steps: {
    test_case_title: string;
    step_number: number;
    action: string;
    error_message?: string;
  }[];
}

export interface PerRunResult {
  test_run_id: string;
  test_case_id: string;
  test_case_title: string;
  status: string;
  duration_ms: number;
  confidence?: number;
  recommendation?: string;
}

export interface FeatureRisk {
  feature_name: string;
  risk_score: number;
  trend_direction: string;
  confidence_level: number;
  pass_rate: number;
  flake_rate: number;
  test_case_count: number;
  total_runs: number;
}

export interface FlakyTest {
  test_case_id: string;
  flake_rate: number;
}

export interface TimingAnomaly {
  test_case_id: string;
  step_number: number;
  action: string;
  stddev_ms: number;
  median_ms: number;
}

export interface TrendPoint {
  id: string;
  score: number | null;
  grade: string | null;
  recommendation: string | null;
  created_at: string;
}

export interface TrendsResponse {
  project_id: string;
  points: TrendPoint[];
}

export interface ReleaseValidationListResponse {
  items: ReleaseValidation[];
  total: number;
  page: number;
  page_size: number;
}
